src/host.rs

```diff
@@ -54,6 +125,72 @@ impl Host {
         }
     }
 
+    pub(crate) fn accept(&mut self) -> Option<Envelope> {
+        let now = Instant::now();
+        let deque = &mut self.listener.as_mut()?.deque;
+
+        match deque.front() {
+            Some(Envelope {
+                instructions: DeliveryInstructions::DeliverAt(time),
+                ..
+            }) if *time <= now => {
+                let ret = deque.pop_front();
+                self.bump_version();
+                ret
+            }
+            _ => None,
+        }
+    }
+
+    // If the host is not bound we simply do nothing, dropping the `Syn`. The
+    // peer who initiated the connection is awaiting the receiver of the syn's
+    // oneshot, which triggers a "connection refused" error.
+    pub(crate) fn syn(&mut self, src: version::Dot, delay: Option<Duration>, syn: Syn) {
+        if let Some(listener) = self.listener.as_mut() {
+            let instructions = match delay {
+                Some(d) => DeliveryInstructions::DeliverAt(self.now + d),
+                None => DeliveryInstructions::ExplicitlyHeld,
+            };
+
+            listener.deque.push_back(Envelope {
+                src,
+                instructions,
+                message: Box::new(syn),
+            });
+
+            listener.notify.notify_one();
+        }
+    }
+
+    /// Finalize the connection, returning a `Notify` to build the stream.
+    ///
+    /// This is idempotent. If the connection has already been initialized, the
+    /// `Notify` is simply cloned and returned.
+    pub(crate) fn finish_connect(&mut self, pair: SocketPair) -> Arc<Notify> {
+        let inbox = self.connections.entry(pair).or_insert_with(|| Inbox {
+            deque: VecDeque::new(),
+            notify: Arc::new(Notify::new()),
+        });
+
+        inbox.notify.clone()
+    }
+
+    pub(crate) fn register_connection(&mut self, pair: SocketPair) {
+        assert!(
+            self.connections
```

Comment:
I don't think this reads very well.

Reply:
Agree, how does this look?

```rust
pub(crate) fn register_connection(&mut self, pair: SocketPair) {
    let inbox = Inbox {
        deque: VecDeque::new(),
        notify: Arc::new(Notify::new()),
    };

    let contains_pair = self.connections.insert(pair, inbox).is_some();

    assert!(!contains_pair, "{:?} is already registered", pair);
}
```

Note that this may change when I clean up `finish_connect()`.